tools/representative_length.py

```diff
@@ -0,0 +1,103 @@
+    """ creates a MemCollection with a line geometry, attributes copied from the input profile file,
+        and three added fields for the representative length: pre length, post length and total
+        representative length.
+
+        receives two line input files:
+        - line shape representing the waterways
+        - line shape representing the profiles
+
+        returns MemCollection profile_col with the representative length data added
+        """
+
+    # Initialising prof_pk; a unique identifier for each profile
+    prof_pk = 0
+
+    # Loop through each line and select the profiles that intersect the line
+    for feature in line_col:
+        # Initialize the collections
+        points_col = MemCollection(geometry_type='Point')
+        sorted_col = MemCollection(geometry_type='Point')
+
+        # Check if line is LineString or MultiLineString and create the appropiate line object
+        if type(feature['geometry']['coordinates'][0][0]) != tuple:
+            line = LineString(feature['geometry']['coordinates'])
+        else:
+            line = MultiLineString(feature['geometry']['coordinates'])
+
+        # Loop through the profiles within the bounding box of the line
+        for profile in profile_col.filter(bbox=line.bounds, precision=10**-6):
+            if type(profile['geometry']['coordinates'][0][0]) != tuple:
+                prof = LineString(profile['geometry']['coordinates'])
+            else:
+                prof = MultiLineString(profile['geometry']['coordinates'])
+
+            # Making an intersection of the profile with the line creates a point
+            x = line.intersection(prof)
+
+            # If the profile has an intersect with the line, and thus a point is created, a unique prof_pk is assigned
+            # to the profile, distance from the beginning of the line to the point is calculated, and it is added to
+            # the points collection.
+            if x:
+                profile['properties']['prof_pk'] = prof_pk
+                prof_pk += 1
+                props = profile['properties'].copy()
+                props['distance'] = line.project(x)
+                points_col.writerecords([
+                    {'geometry': {'type': 'Point',
+                                  'coordinates': x.coords[0]},
+                     'properties': props}]
+                )
+
+        # After all profiles on a line are defined, sort them by distance so that they are in order
+        sorted_list = sorted(points_col.keys(), key=lambda x: (points_col[x]['properties']['distance']))
+
+        # Write the sorted points to another collection (sorted_col)
+        for i in sorted_list:
+            p = points_col[i]
+            props = p['properties']
+            coords = p['geometry']['coordinates']
+
+            sorted_col.writerecords([
+                {'geometry': {'type': 'Point',
+                              'coordinates': coords},
+                 'properties': props}])
```

Comment:
bovengaande ziet er wat omslachtig uit, volgens mij kan dit netter en korter.

Reply:
Als dat kan dan hoor ik dat graag. Ik heb deze methode namelijk ook al toegepast in 'load_veldwerk.py'. Een dictionary kan niet in-place gesorteerd worden en uit de sorteerfunctie ontstaat alleen een list met indices.